lib/charon/session_store/local_store.ex

```diff
@@ -0,0 +1,110 @@
+defmodule Charon.SessionStore.LocalStore do
+  @moduledoc """
+  An in-memory persistent session store, implements behaviour `Charon.SessionStore`.
+  In addition to the required callbacks, this store also provides `get_all/3` and `delete_all/3` (for a user) functions.
+
+  Add this store to your supervision tree to use it.
+  """
+  use Agent
+
+  @behaviour Charon.SessionStore.Behaviour
+  alias Charon.Internal
+  alias Charon.Models.Session
+  require Logger
+
+  def start_link(_) do
+    Agent.start_link(fn -> %{} end, name: __MODULE__)
+  end
+
+  @impl true
+  def get(session_id, user_id, type, _config) do
+    Agent.get(__MODULE__, fn state ->
+      session = Map.get(state, {session_id, user_id, type}, nil)
+
+      if session == nil || Internal.now() > session.refresh_expires_at do
+        nil
+      else
+        session
+      end
+    end)
+  end
+
+  @impl true
+  def upsert(
+        session = %{
+          id: session_id,
+          user_id: user_id,
+          type: type
+        },
+        _config
+      ) do
+    Agent.update(__MODULE__, fn state ->
+      Map.put(state, {session_id, user_id, type}, session)
+    end)
+  end
+
+  @impl true
+  def delete(session_id, user_id, type, _config) do
+    Agent.update(__MODULE__, fn state ->
+      Map.drop(state, [{session_id, user_id, type}])
+    end)
+  end
+
+  @impl true
+  def get_all(user_id, type, _config) do
+    now = Internal.now()
+
+    Agent.get(__MODULE__, fn state ->
+      state
+      |> Stream.filter(fn
+        {{_, ^user_id, ^type}, %{refresh_expires_at: exp}} when exp > now -> true
+        _ -> false
+      end)
+      |> Stream.map(fn {_, v} -> v end)
+      |> Enum.to_list()
+    end)
+  end
+
+  @impl true
+  def delete_all(user_id, type, _config) do
+    Agent.update(__MODULE__, fn state ->
+      user_session_keys =
+        state
+        |> Map.keys()
+        |> Enum.filter(fn {_, state_uid, state_type} ->
+          {user_id, type} == {state_uid, state_type}
+        end)
+
+      Map.drop(state, user_session_keys)
```

Comment:
This loops over the state three times, while it can be done in a single pass using `Map.reject/2`, can you give it another go?